// @twurple/api/lib/api/helix/goals/HelixGoalApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixGoalApi = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
const BaseApi_1 = require("../../BaseApi");
const HelixGoal_1 = require("./HelixGoal");
/**
 * The Helix API methods that deal with creator goals.
 *
 * Can be accessed using `client.helix.goals` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const { data: goals } = await api.helix.goals.getGoals('61369223');
 */
let HelixGoalApi = class HelixGoalApi extends BaseApi_1.BaseApi {
    async getGoals(broadcaster) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'goals',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(broadcaster)
            }
        });
        return result.data.map(data => new HelixGoal_1.HelixGoal(data, this._client));
    }
};
HelixGoalApi = tslib_1.__decorate([
    (0, common_1.rtfm)('api', 'HelixGoalApi')
], HelixGoalApi);
exports.HelixGoalApi = HelixGoalApi;
