// @twurple/api/lib/api/helix/moderation/HelixModerationApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixModerationApi = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
const BaseApi_1 = require("../../BaseApi");
const HelixPaginatedRequest_1 = require("../HelixPaginatedRequest");
const HelixPaginatedResult_1 = require("../HelixPaginatedResult");
const HelixPagination_1 = require("../HelixPagination");
const HelixAutoModSettings_1 = require("./HelixAutoModSettings");
const HelixAutoModStatus_1 = require("./HelixAutoModStatus");
const HelixBan_1 = require("./HelixBan");
const HelixModerator_1 = require("./HelixModerator");
/**
 * The Helix API methods that deal with moderation.
 *
 * Can be accessed using `client.moderation` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const { data: users } = await api.moderation.getBannedUsers('61369223');
 * ```
 */
let HelixModerationApi = class HelixModerationApi extends BaseApi_1.BaseApi {
    /**
     * Retrieves a list of banned users in a given channel.
     *
     * @param channel The channel to retrieve the banned users from.
     * @param filter Additional filters for the result set.
     */
    async getBannedUsers(channel, filter) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/banned',
            scope: 'moderation:read',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(channel),
                user_id: filter === null || filter === void 0 ? void 0 : filter.userId,
                ...(0, HelixPagination_1.makePaginationQuery)(filter)
            }
        });
        return (0, HelixPaginatedResult_1.createPaginatedResult)(result, HelixBan_1.HelixBan, this._client);
    }
    /**
     * Creates a paginator for banned users in a given channel.
     *
     * @param channel The channel to retrieve the banned users from.
     */
    getBannedUsersPaginated(channel) {
        return new HelixPaginatedRequest_1.HelixPaginatedRequest({
            url: 'moderation/banned',
            scope: 'moderation:read',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(channel)
            }
        }, this._client, data => new HelixBan_1.HelixBan(data, this._client), 50 // possibly a relatively consistent workaround for twitchdev/issues#18
        );
    }
    /**
     * Checks whether a given user is banned in a given channel.
     *
     * @param channel The channel to check for a ban of the given user.
     * @param user The user to check for a ban in the given channel.
     */
    async checkUserBan(channel, user) {
        const userId = (0, common_1.extractUserId)(user);
        const result = await this.getBannedUsers(channel, { userId });
        return result.data.some(ban => ban.userId === userId);
    }
    /**
     * Retrieves a list of moderators in a given channel.
     *
     * @param channel The channel to retrieve moderators from.
     * @param filter Additional filters for the result set.
     */
    async getModerators(channel, filter) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/moderators',
            scope: 'moderation:read',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(channel),
                user_id: filter === null || filter === void 0 ? void 0 : filter.userId,
                ...(0, HelixPagination_1.makePaginationQuery)(filter)
            }
        });
        return (0, HelixPaginatedResult_1.createPaginatedResult)(result, HelixModerator_1.HelixModerator, this._client);
    }
    /**
     * Creates a paginator for moderators in a given channel.
     *
     * @param channel The channel to retrieve moderators from.
     */
    getModeratorsPaginated(channel) {
        return new HelixPaginatedRequest_1.HelixPaginatedRequest({
            url: 'moderation/moderators',
            scope: 'moderation:read',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(channel)
            }
        }, this._client, data => new HelixModerator_1.HelixModerator(data, this._client));
    }
    /**
     * Checks whether a given user is a moderator of a given channel.
     *
     * @param channel The channel to check.
     * @param user The user to check.
     */
    async checkUserMod(channel, user) {
        const userId = (0, common_1.extractUserId)(user);
        const result = await this.getModerators(channel, { userId });
        return result.data.some(mod => mod.userId === userId);
    }
    /**
     * Determines whether a string message meets the channel's AutoMod requirements.
     *
     * @param channel The channel in which the messages to check are posted.
     * @param data An array of message data objects.
     */
    async checkAutoModStatus(channel, data) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/enforcements/status',
            method: 'POST',
            scope: 'moderation:read',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(channel)
            },
            jsonBody: {
                data: data
            }
        });
        return result.data.map(statusData => new HelixAutoModStatus_1.HelixAutoModStatus(statusData));
    }
    /**
     * Processes a message held by AutoMod.
     *
     * @param user The user who is processing the message.
     * @param msgId The ID of the message.
     * @param allow Whether to allow the message - `true` allows, and `false` denies.
     */
    async processHeldAutoModMessage(user, msgId, allow) {
        await this._client.callApi({
            type: 'helix',
            url: 'moderation/automod/message',
            method: 'POST',
            scope: 'moderator:manage:automod',
            jsonBody: {
                user_id: (0, common_1.extractUserId)(user),
                msg_id: msgId,
                action: allow ? 'ALLOW' : 'DENY'
            }
        });
    }
    /**
     * Retrieves the AutoMod settings for a broadcaster.
     *
     * @param broadcasterId The ID of the broadcaster for which the AutoMod settings are retrieved.
     * @param moderatorId The ID of a user that has permission to moderate the broadcaster's chat room.
     * This must match the user ID associated with the user OAuth token.
     */
    async getAutoModSettings(broadcasterId, moderatorId) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/automod/settings',
            scope: 'moderator:read:automod_settings',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(broadcasterId),
                moderator_id: (0, common_1.extractUserId)(moderatorId)
            }
        });
        return result.data.map(data => new HelixAutoModSettings_1.HelixAutoModSettings(data));
    }
    /**
     * Updates the AutoMod settings for a broadcaster.
     *
     * @param broadcasterId The ID of the broadcaster for which the AutoMod settings are updated.
     * @param moderatorId The ID of a user that has permission to moderate the broadcaster's chat room.
     * This must match the user ID associated with the user OAuth token.
     * @param data The updated AutoMod settings that replace the current AutoMod settings.
     */
    async updateAutoModSettings(broadcasterId, moderatorId, data) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/automod/settings',
            method: 'PUT',
            scope: 'moderator:manage:automod_settings',
            query: {
                broadcaster_id: (0, common_1.extractUserId)(broadcasterId),
                moderator_id: (0, common_1.extractUserId)(moderatorId)
            },
            jsonBody: {
                overall_level: data.overallLevel,
                aggression: data.aggression,
                bullying: data.bullying,
                disability: data.disability,
                misogyny: data.misogyny,
                race_ethnicity_or_religion: data.raceEthnicityOrReligion,
                sex_based_terms: data.sexBasedTerms,
                sexuality_sex_or_gender: data.sexualitySexOrGender,
                swearing: data.swearing
            }
        });
        return result.data.map(settingsData => new HelixAutoModSettings_1.HelixAutoModSettings(settingsData));
    }
};
HelixModerationApi = tslib_1.__decorate([
    (0, common_1.rtfm)('api', 'HelixModerationApi')
], HelixModerationApi);
exports.HelixModerationApi = HelixModerationApi;
